security/fips: add fips guide (#251)
flatcar-linux/flatcar-docs: docs/setup/security/fips.md

---
title: Flatcar Container Linux FIPS guide
linktitle: FIPS mode
description: Enabling FIPS mode.
weight: 20
---

> FIPS stands for Federal Information Processing Standards, a set of standards issued by the National Institute of Standards and Technology (NIST). With Flatcar, it's possible to have a compliant instance with two of these standards:

Reviewer: ```suggestion
FIPS stands for Federal Information Processing Standards, a set of standards issued by the National Institute of Standards and Technology (NIST). While Flatcar is not officially FIPS certified, it is possible to deploy it so that it is compliant with two of these standards:
```

* [FIPS 200][fips-200]
* [FIPS 140-2][fips-140-2]

# Enabling FIPS

> Booting the instance with `fips=1` allows the instance to operate in a FIPS 200 certification: Kernel will use FIPS compliant algorithms and will enforce some security practices like RSA key [size][rsa-key-size].

Reviewer: ```suggestion
Booting the instance with the kernel parameter `fips=1` allows the instance to operate in a FIPS 200 mode. This means the kernel will use FIPS-compliant algorithms and will enforce some security practices like RSA key [size][rsa-key-size].
```

> Booting the instance with `fips=1` allows the instance to operate in a FIPS 200 certification: Kernel will use FIPS compliant algorithms and will enforce some security practices like RSA key [size][rsa-key-size].

Reviewer 2: As far as I know we have FIPS also enabled in the cryptsetup tools, or? Would be good to tell how this is related to the kernel setting, e.g., does the kernel only accept LUKS devices created with Flatcar's cryptsetup tools or will it only be LUKS compliant if the LUKS device is created with Flatcar's cryptsetup tools?

Author: Thanks for the cryptsetup heads-up. I missed the `/etc/system-fips` for cryptsetup - otherwise it fallbacks on the default "legacy" OpenSSL provider.
I'm first to create an encrypted device with cryptsetup + FIPS but for now it's failling:
```
sudo cryptsetup luksFormat ./volume
Not compatible PBKDF2 options (using hash algorithm sha256).
``

Author: > does the kernel only accept LUKS devices created with Flatcar's cryptsetup tools or will it only be LUKS compliant if the LUKS device is created with Flatcar's cryptsetup tools?

I created a LUKS device on my personal machine - it can't be opened on Flatcar with FIPS. Because without FIPS, cryptsetup will use `argon2id` by default to run key derivation - but this algorithm is not FIPS compliant, that's why it does not work.
OTOH if one specifies `--pbkdf=pbkdf2`, even from a non-fips booted machine it will work on the FIPS machine.

Reviewer 2: Thanks for checking, this needs to be part of the docs.

> It's possible to check that FIPS mode is enabled on the Kernel by checking the content of this file `/proc/sys/crypto/fips_enabled`:

Reviewer: ```suggestion
To confirm that FIPS mode is enabled on the Kernel, check the content of the file `/proc/sys/crypto/fips_enabled`:
```

```bash
$ cat /proc/sys/crypto/fips_enabled
0 # disabled
1 # enabled
```

or by inspecting boot logs:
```bash
$ journalctl | grep -i "kernel: fips"
Jun 27 18:07:22 localhost kernel: fips mode: enabled
```

# Enabling OpenSSL FIPS provider

> [OpenSSL][openssl] is an open-source library used for ciphering and hashing. As a library, it is widely used by programming software and third-party programs to ensure security. OpenSSL has [submitted][openssl-submission] its FIPS module for being reviewed and validated by the NIST’s Cryptographic Module Validation Program. It’s important to keep in mind that OpenSSL is FIPS compliant but not yet FIPS validated.

Reviewer: ```suggestion
[OpenSSL][openssl] is an open-source library used for ciphering and hashing. As a library, it is widely used by programming software and third-party programs to ensure security. OpenSSL has [submitted][openssl-submission] its FIPS module for review and validation by the NIST’s Cryptographic Module Validation Program. Hence, currently OpenSSL is FIPS compliant but not yet FIPS validated.
```

> OpenSSL FIPS module is built by default on Flatcar but it's required to update the OpenSSL configuration to actually use this module:

Reviewer: ```suggestion
OpenSSL FIPS module is built by default on Flatcar but it is required to update the OpenSSL configuration to actually use this module:
```

```bash
openssl fipsinstall \
    -out /etc/ssl/fipsmodule.cnf \
    -module /usr/lib64/ossl-modules/fips.so
mv /etc/ssl/openssl.cnf.fips /etc/ssl/openssl.cnf
```

Once again, it's possible to check that FIPS is enabled:
```bash
$ echo "Flatcar + FIPS" | openssl sha1 -
SHA1(stdin)= ee2219bd6a234fa0e4436b475fc3b351e2dc85a0
$ echo "Flatcar + FIPS" | openssl md5 -
Error setting digest C0422ACDB57F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (MD5 : 104), Properties ()C0422ACDB57F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:252:
```

# Ignition provisioning

> It's possible to wrap-up these two sections into one Ignition configuration:

Reviewer: ```suggestion
The two sections above can be combined into one Ignition configuration, as follows.
```

          # by the `enable-fips.service`
          .include /etc/ssl/fipsmodule.cnf
          [openssl_init]
          providers = provider_sect
          [provider_sect]
          fips = fips_sect
          base = base_sect
          [base_sect]
          activate = 1
systemd:
  units:
    - name: enable-fips.service
      enabled: true
      contents: |
        [Unit]
        Description=Enable OpenSSL FIPS provider
        ConditionPathExists=!/etc/ssl/fipsmodule.cnf
        After=system-config.target
        [Service]
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/usr/bin/openssl fipsinstall \
          -out /etc/ssl/fipsmodule.cnf \
          -module /usr/lib64/ossl-modules/fips.so
        ExecStart=/usr/bin/mv /etc/ssl/openssl.cnf.fips /etc/ssl/openssl.cnf
        [Install]
        WantedBy=multi-user.target
```

# Troubleshooting

## SSH login does not work with OpenSSL FIPS provider

It's possible to have a SSH connection refused when OpenSSL FIPS provider is enabled. Inspecting the SSHd logs:
```bash
Jun 28 07:58:39 localhost sshd[1080]: ssh_dispatch_run_fatal: Connection from 10.0.2.2 port 40192: invalid argument [preauth]
```

> It might be one of the `Ciphers`, defined in the `/etc/ssh/sshd_config`, that is not FIPS compliant (like `chacha20-poly1305`).

Reviewer: ```suggestion
In this case, it is likely that one of the `Ciphers`, defined in the `/etc/ssh/sshd_config`, is not FIPS compliant (like `chacha20-poly1305`).
```

---
title: Flatcar Container Linux FIPS guide
linktitle: FIPS mode
description: Enabling FIPS mode.
weight: 20
---

FIPS stands for Federal Information Processing Standards, a set of standards issued by the National Institute of Standards and Technology (NIST). While Flatcar is not officially FIPS certified, it is possible to deploy it so that it is compliant with two of these standards:
* [FIPS 200][fips-200]
* [FIPS 140-2][fips-140-2]

# Enabling FIPS

> Booting the instance with the kernel parameter `fips=1` allows the instance to operate in a FIPS 200 mode. This means the kernel will use FIPS-compliant algorithms and will enforce some security practices like RSA key [size][rsa-key-size].

Reviewer 3: having `fips=1` does boot the kernel with these parameters, but additionally, the libraries behind `fipscheck` also look to see if the host filesystem has ` /etc/system-fips`. It's a little silly, but it does check. The file can be zero length, so just ` touch /etc/system-fips` will be enough.

Reviewer 3: Maybe it is the case that it is not _always_ checked for. I need to find a source calling out why it was ever checked ...

Reviewer 3: https://github.com/search?q=%22%2Fetc%2Fsystem-fips%22&type=code

Author: Yeah it's also required by cryptsetup: https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/utils_fips.c#L50 - I added this file to the Butane config in the latest commit.

To confirm that FIPS mode is enabled on the Kernel, check the content of the file `/proc/sys/crypto/fips_enabled`:
```bash
$ cat /proc/sys/crypto/fips_enabled
0 # disabled
1 # enabled
```

or by inspecting boot logs:
```bash

> $ journalctl | grep -i "kernel: fips"

Reviewer 4: I'd do `journalctl --boot` to show if currently booted kernel has fips mode enabled.

OpenSSL FIPS module is built by default on Flatcar but it is required to update the OpenSSL configuration to actually use this module:
```bash
openssl fipsinstall \
    -out /etc/ssl/fipsmodule.cnf \
    -module /usr/lib64/ossl-modules/fips.so
mv /etc/ssl/openssl.cnf.fips /etc/ssl/openssl.cnf
```

Once again, it's possible to check that FIPS is enabled:
```bash
$ echo "Flatcar + FIPS" | openssl sha1 -
SHA1(stdin)= ee2219bd6a234fa0e4436b475fc3b351e2dc85a0
$ echo "Flatcar + FIPS" | openssl md5 -
Error setting digest C0422ACDB57F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (MD5 : 104), Properties ()C0422ACDB57F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:252:
```

OpenSSL FIPS module is also being used by `cryptsetup` when running in FIPS mode (detection is based on `fips` kernel parameter and `/etc/system-fips` file).

To check that cryptsetup runs in FIPS mode, it's possible to add the `--verbose` flag:
```bash
$ cryptsetup --verbose luksFormat ./volume
...
Running in FIPS mode.
Command successful.
```

> _NOTE_: Formatting a LUKS device with `cryptsetup` on a non-FIPS instance will use `argon2id` as key derivation function. This algorithm is not FIPS-compliant, so it will be impossible to open the LUKS device on a FIPS instance. It is possible to have a FIPS-compatible LUKS device if it is formatted using `cryptsetup luksFormat --pbkdf=pbkdf2 ./my-volume`.

Reviewer 2: ```suggestion
_NOTE_: Formatting a LUKS device with `cryptsetup` on a non-FIPS instance will use `argon2id` as key derivation function. This algorithm is not FIPS-compliant, so it will be impossible to open the LUKS device on a FIPS instance. It is possible to have a FIPS-compatible LUKS device if it is formatted using `cryptsetup luksFormat --pbkdf=pbkdf2 ./my-volume` which is the default behavior on Flatcar even if `--pbkdf=pbkdf2` is not specified.
```

Is this clarification correct?

Author: Mm, may I propose this:
> which is the default behavior on a Flatcar FIPS instance even if `--pbkdf=pbkdf2` is not specified

Reviewer 2: Thanks, yes, this would be depending on the flag file, right? I've updated it

Author: Yes. TL; DR:
* FIPS instance default to `--pbkdf=pbkdf2`
* non-FIPS instance default to `--pbkdf=argon2id`

FIPS:
```
$ cat /proc/sys/crypto/fips_enabled
1
$ cryptsetup --help | grep "Default PBKDF"
Default PBKDF for LUKS1: pbkdf2, iteration time: 2000 (ms)
Default PBKDF for LUKS2: pbkdf2
```

Non-FIPS:
```
cryptsetup --help | grep "Default PBKDF"
Default PBKDF for LUKS1: pbkdf2, iteration time: 2000 (ms)
Default PBKDF for LUKS2: argon2id
```
